graspologic/pipeline/embed/embeddings.py

```diff
@@ -0,0 +1,122 @@
+"""
+Embeddings
+==========
+
+The ``Embeddings`` and ``EmbeddingsView`` classes combine the 2 dimensional
+embeddings generated for a graph and the associated node labels for each embedding.
+
+The embeddings generated follow the natural ordering of the provided networkx Graph
+object, but as they are positionally correlated and the source graph object may diverge
+in that natural ordering due to mutations, it is often helpful to combine a snapshot
+of the labels at the same time the embedding was genrated with the embedding to
+guarantee the appropriate latent positions are returned for each node.
+"""
+
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+
+from typing import Any, Tuple
+import numpy as np
+from collections import OrderedDict
+
+
+class Embeddings:
+    def __init__(self, labels: np.ndarray, embeddings: np.ndarray):
+        """
+        ``Embeddings`` is an iterable, indexed interface over the parallel numpy arrays
+        that are generated by our embedding functions.
+
+        >>> import numpy as np
+        >>> raw_embeddings = np.array([[1,2,3,4,5], [6,4,2,0,8]])
+        >>> labels = np.array(["monotonic", "not_monotonic"])
+        >>> embeddings_obj = Embeddings(labels, raw_embeddings)
+        >>> embeddings_obj[0]
+        ('monotonic', array([1, 2, 3, 4, 5]))
+        >>> len(embeddings_obj)
+        2
+        >>> list(iter(embeddings_obj))
+        [('monotonic', array([1, 2, 3, 4, 5])), ('not_monotonic', array([6, 4, 2, 0, 8]))]
+        >>> embeddings_lookup = embeddings_obj.as_dict()
+        >>> embeddings_lookup["not_monotonic"]
+        array([6, 4, 2, 0, 8])
+
+        Parameters
+        ----------
+        labels : np.ndarray
+            The node labels that are positionally correlated with the embeddings.
+            The dtype of labels is any object stored in a networkx Graph object,
+            though type uniformity will be required
+        embeddings
+        """
+        if labels is None:
+            raise ValueError("labels cannot be None")
+        if embeddings is None:
+            raise ValueError("embeddings cannot be None")
+        if not isinstance(labels, np.ndarray):
+            raise TypeError(f"labels must be numpy.ndarray, got: {type(labels)}")
+        if not isinstance(embeddings, np.ndarray):
+            raise TypeError(
+                f"embeddings must be numpy.ndarray, got: {type(embeddings)}"
+            )
+        if labels.shape[0] != embeddings.shape[0]:
+            raise ValueError(
+                f"labels and embeddings must have the same number of "
+                f"rows, labels: {labels.shape[0]}, "
+                f"embeddings: {embeddings.shape[0]}"
+            )
+        self._labels = labels
+        self._embeddings = embeddings
+
+    def labels(self) -> np.ndarray:
+        return self._labels
+
+    def embeddings(self) -> np.ndarray:
+        return self._embeddings
```

Comment:
would these make sense as properties?

Reply:
I didn't really want people to be thinking it made sense to set these at any point.  we could but I donno that it buys us anything important really?

Reply:
i guess just whether you access them by obj.embeddings vs obj.embeddings() (method vs attribute-like thing) was my main point

Reply:
So it's basically a subtle semantics implication/problem and me trying to do my best to limit unsupported/unintended usage:
`obj.embeddings` -> as an attribute, you can access it directly, safely, and mutate it without regard for side effects (which is a terrible idea in this case)
`obj._embeddings` -> as an attribute, it's private: you are risking the Embedding integrity by mutating it, and you really shouldn't look at it
`obj._embeddings` with `obj.embeddings()` getter function: you can safely get these values to look at, but probably only through this method

Now, I know we're just passing a pointer in the latter case, and you can mutate it all you want - and you can mutate the attribute all you want too.  Python lets you do whatever you want whenever you want without guard rails.

Since the language doesn't let us limit a lot of behaviors that other languages do, we're stuck defining things that *imply* the behaviors we're wanting to support and hoping people agree to observe the rules of what is implied and don't try to ... well, do whatever crazy stuff they want to do.